packages/ai-chat/src/common/chat-parsed-request.ts

```diff
@@ -0,0 +1,135 @@
+export const chatAgentLeader = '@';
+export const chatFunctionLeader = '~';
+export const chatSubcommandLeader = '/';
+
+/**********************
+ * INTERFACES AND TYPE GUARDS
+ **********************/
+
+export interface OffsetRange {
+    readonly start: number;
+    readonly endExclusive: number;
+}
+export class OffsetRangeImpl implements OffsetRange {
+    constructor(public readonly start: number, public readonly endExclusive: number) {
+        if (start > endExclusive) {
+            throw new Error(`Invalid range: ${this.toString()}`);
+        }
+    }
+}
+
+export interface ParsedChatRequest {
+    readonly request: ChatRequest;
+    readonly parts: ParsedChatRequestPart[];
+    readonly toolRequests: Map<string, ToolRequest<object>>;
+    readonly variables: Map<string, AIVariable>;
+}
+
+export interface ChatRequestBasePart {
+    readonly kind: string;
+    /**
+     * The text as represented in the ChatRequest
+     */
+    readonly text: string;
+    /**
+     * The text as will be sent to the LLM
+     */
+    readonly promptText: string;
+
+    readonly range: OffsetRange;
+}
+
+export class ChatRequestTextPart implements ChatRequestBasePart {
+    readonly kind: 'text';
+
+    constructor(readonly range: OffsetRange, readonly text: string) { }
+
+    get promptText(): string {
+        return this.text;
+    }
+}
+
+export class ChatRequestVariablePart implements ChatRequestBasePart {
+    readonly kind: 'var';
+
+    protected _resolution: ResolvedAIVariable;
+
+    constructor(readonly range: OffsetRange, readonly variableName: string, readonly variableArg: string | undefined) { }
+
+    get text(): string {
+        const argPart = this.variableArg ? `:${this.variableArg}` : '';
+        return `${chatVariableLeader}${this.variableName}${argPart}`;
+    }
+
+    get promptText(): string {
+        return this._resolution?.value ?? this.text;
+    }
+
+    resolve(resolution: ResolvedAIVariable): void {
+        this._resolution = resolution;
+    }
+
+    get resolution(): ResolvedAIVariable | undefined {
+        return this._resolution;
+    }
+}
+
+export class ChatRequestFunctionPart implements ChatRequestBasePart {
+    readonly kind: 'function';
+    constructor(readonly range: OffsetRange, readonly toolRequest: ToolRequest<object>) { }
+
+    get text(): string {
+        return `${chatFunctionLeader}${this.toolRequest.id}`;
+    }
+
+    get promptText(): string {
+        return toolRequestToPromptText(this.toolRequest);
+    }
+}
+
+export class ChatRequestAgentPart implements ChatRequestBasePart {
+    readonly kind: 'agent';
+    constructor(readonly range: OffsetRange, readonly agent: ChatAgentData) { }
+
+    get text(): string {
+        return `${chatAgentLeader}${this.agent.name}`;
+    }
+
+    get promptText(): string {
+        return '';
+    }
+}
+
+export type ParsedChatRequestPart = ChatRequestBasePart | ChatRequestTextPart | ChatRequestVariablePart | ChatRequestAgentPart;
```

Comment:
They don't have the same interface. The differences are defined in their constructors which are implicitly defining additional fields.

By constructing the `ParsedChatRequestPart` type this way, Typescript will be able to differentiate these fields just by checking the `kind` field.

Reply:
I don't think that's the case: what is keeping me from writing 
```
const part: ParsedChatRequestPart = {
   kind: ''var'',
   text: 'foobar',
   promptText: 'prompt',
   rang: { 5, 7 }
}
```

This is a perfectly valid `ChatRequestBasePart`, but not a `ChatRequestVariablePart`, despite the kind field.

Reply:
You are right, because we include the `BaseChatRequestPart` in the union it does not really gain us any benefits. Our type guards check all properties already anyway, so I'm fine with removing the union here.

Reply:
I'll remove it